### ISSUE TITLE
[MNG-7001] Reconsider seemingly useless check of artifacts' source repository introduced in Maven 3.0

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8031]
**[Petr Bodnar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pbodnar)** opened **[MNG-7001](https://issues.apache.org/jira/browse/MNG-7001?redirect=false)** and commented

This problem of "by-nobody-really-requested check for artifacts' source repository" (just "repo-check" further on) is actually considered a bug by many Maven users. It was introduced back in Maven 3.0, 10 years ago \(!). The repo-check and its _practical_ disadvantages have been already thoroughly described for example in my blog [here](https://programmedbycoincidence.blogspot.com/2019/01/the-biggest-wtf-new-feature-ive-ever.html) and discussed here within Jira: MNG-5181, MNG-5185, MNG-5289 and MNG-5883.

**TL;DR What is requested in this issue:**

1. Remove the repo-check altogether.
2. If that's not possible, make the repo-check disabled by-default and have an option to enable it for those who need it for whatever reason.
3. If even that is not possible, alter Maven and its warnings and errors so that they do not confuse users.
4. Reason about the need for the repo-check, document the reasons.

---

The repo-check can be _somewhat_ avoided by passing the `-llr` option to Maven. AFAIK though, e. g. Eclipse's embedded Maven used for dependency resolution doesn't support this option. Another long-outstanding issue is that using the `-llr` option generates this warning on Maven build:

```
[WARNING] Disabling enhanced local repository: using legacy is strongly discouraged to ensure build reproducibility.
```

Generally it might make sense (possibly because of activating some quite another old part of Maven that, apart from other things, doesn't mark down the artifacts' sources to "\*.repositories" files?). But when users have _no other option_ that could be used for making their build reproducible by skipping the repo-check, then the warning doesn't make sense to them. The only other choice they have is to remove all those "\*.repositories" files from their local Maven repository in order to make their builds work again.

Another mind-blowing issue is described in MNG-5185: If an already-downloaded artifact doesn't go through the hard-coded repo-check, Maven just tells the user "the artifact could not be resolved". _But you'll get the very same message when downloading an artifact really fails._ So unless you dig in, these two totally different situations are not distinguishable from each other.

---

Yet to date, no action was taken by Maven authors to help with any of the problems. There is also no really good (read "making-sense-in-real-life") explanation of real pros of the introduced repo-check, that would out-weight its cons, other than for example:

> The artifacts have an identity. It matters where the artifacts were downloaded from. Artifact A downloaded from X is not the same thing to Maven 3 as A downloaded from Y. This can happen when you flip your settings.xml to go from using a repository manager to using Maven Central directly for example.

(taken from MNG-5289 comment)

The logical question here is, to whom concretely "it matters"? Please, give examples of what could go wrong if one has downloaded a released version of an artifact and now its source repository changes or becomes unavailable.

Please note that we shouldn't consider the very improbable case of artifacts downloaded from various repositories would have different content even though having the very same GAV. The Maven's local repository filesystem structure is not able to cope with that situation anyway, or is it?

Finally, there is also a performance-wise con of the repo-check - Maven needs to contact the source repository every time it builds a project referencing the checked artifact as one of its dependencies. Or doesn't it?

---

**Affects:** 3.0, 3.1.1, 3.2.5, 3.3.9, 3.5.4, 3.6.3

**Issue Links:**
- [MDEP-569](https://issues.apache.org/jira/browse/MDEP-569) copy fails on artifacts previously downloaded via get from repo specified via CLI

- [MNG-6889](https://issues.apache.org/jira/browse/MNG-6889) Mark command line option --legacy-local-repository as deprecated

- [MNG-5289](https://issues.apache.org/jira/browse/MNG-5289) -Dmaven.repo.local not honored

- [MNG-5181](https://issues.apache.org/jira/browse/MNG-5181) New resolution from local repository is very confusing

- [MNG-5883](https://issues.apache.org/jira/browse/MNG-5883) Silence unnecessary legacy local repository warning

- [MNG-5185](https://issues.apache.org/jira/browse/MNG-5185) Improve "missing dependency" error message when _maven.repositories/_remote.repositories contains other repository ids than requested
 (_**"is superceded by"**_)

2 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

The check is still correct, but it is too simple. What should be done if verify if the hashes/checksums are the same. In that case it shouldn't matter if they come from a different repository. 

What would help, is if people provide example projects that rely on this flag, so we understand how to improve the implementation.

**Comment by jira-importer** [comment]
**[Petr Bodnar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pbodnar)** commented

> We must also bear in mind that we have very little resources, rewriting essential parts of Resolver to sastify a too simple local repo manager is just not feasible for us. We must find a balance between reality and correctness. A simple example: I have spends hours on a simple patch for MNG-6754 because they is a lot of code spots and verifications, yet I don't have an IT for it. How many weeks of discussion, design, implementation and testing we'll spend to make it less simple.

Well, hard to tell for me...

I think this is **what needs to be technically done to match the real-world scenarios** as claimed by me and others, in short:
1. Find the code that does the repo-check. Hopefully, it will be just at one place.
2. In order to fix the error message:
   1. If the repo-check code generates a too general exception on check failure, change it so that it throws a specific exception.
   2. At the Maven appropriate level, identify the specific exception and offer a new understandable message accordingly.
3. In order to give users control over activating the repo-check:
   1. Introduce new Maven switches on appropriate levels to enable / disable the repo-check. This could be settings.xml, CLI switch and/or a parameter passed to a corresponding API call (like when an embedded Maven is called from Eclipse IDE directly). (a deeper analysis probably needed)
   2. Propagate the resulting boolean value of the new setting down to the repo-check code (or a step above).
   3. Execute the repo-check code only when the setting enables the repo-check, i. e. wrap the existing code into an IF block.
   4. (As a side-effect of implementing this, the "-llr" option would probably no longer be needed and it could be removed from the code?)

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), how much work do you expect that this would take? And how much could it take for a "newbie" like me if he decides to investigate and try to alter the code himself (including Maven + Maven Resolver development environment setup)? Even though the changes seem like nothing extra big, should we expect complications, as might be possibly implied from what you write?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

> Interesting idea. Is it possible to find out the proxied repo purely on the client side (i. e. in Maven), or would there also be a need to change the server side (i. e. Nexus)? As I said, I'm no expert, but I guess that this could be for a separate issue provided it is a viable idea?

Since repo groups emerged at some point we need to stop treading mirrors like repos, we need to treat them as we treat HTTP proxies: fully transparent. So not the mirror id needs to be persistent, but the original one. This cannot be solely achieved on the client side. The repo manager has to tell me this. But here are there problems:
* Who will change all mirror implementations
* How to transport metadata transport-agnostic. SFTP does not offer such things as HTTP headers, etc.

Let's get back to your points:

1. This one is easy.  I have found this already. It's directly in the enhanced local repo manager impl.
2. The exception is generated with Maven Resolver, it needs to be addressed there.

It hard to tell how long this is going to take until we agree on something as a starting point. Maven Resolver code is quite straight forward to understand. Moreover, will an updated appraoch require changes in repo managers, settings model, Maven itself or can we solely solve it in Resolver?!

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

IIRC [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) would agree that repository managers should be handled as a proxy, not as a mirror. Looking back, the wildcard in mirror to refer to a repository manager was a poor choice.

IMO the repo-check should be based on the online-mode (i.e. don't check in offline mode). If that's not enough we need to figure out why there would be a need to control the check and solve that.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

And just ftr, "repo groups" were a bloody mistake in Proximity1, carried over with offspring MRMs :)

**Comment by jira-importer** [comment]
**[Petr Bodnar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pbodnar)** commented

Thank you guys for taking this further.

So regarding the repo-check, for a start that shouldn't take too much time, I would suggest:
1. To change the error repo-check failure message - this is already covered by MNG-5185 and should be implemented there, together with completing the http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException page. From what was written by [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), it seems to be trivial to do? The hardest part will be possibly to come up with a good error message (that authors of the repo-check should have supplied years ago already).
2. To skip the repo-check at least when in offline mode (for "-o" and possibly also for "-nsu"). As this will not change settings model etc., this could be pretty easy to do. As you guys said, anything else would be too time consuming to be realized in a reasonable time now.

What do you think?

> If that's not enough we need to figure out why there would be a need to control the check and solve that.

Yes, even though the above gets implemented, I expect complaints from Maven users will still come. Because together with the "-llr" option, it is just a _partial_ solution to the problems described. Hopefully some other guys will come and they will be able to re-explain it to you. I'm already a bit tired from discussing this when the only argument is that "it solves our problem and if you want to do it differently, you are doing it wrong"...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas), looking at the problems now, I argree with you 100%.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Petr Bodnar](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pbodnar),

1. Makes sense, find the spot and think about a good message. Yes, this is hard ;)
2. This needs to be discussed as a (interims) solution.

I am includes to say that the repo check must only be valid for repos and **not** for mirrors because you don't know whether a mirror a 1:1 replacement or a repo group.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

A very easy check, is Maven or user "wrong", is simply to **nuke your local repository and rebuild** your project.

In short, a project that FAILS with artifact present in local repository (LR) as being unable to resolve it, will also fail with empty local repository as well. That's all in short.

Keep in mind, that LR is merely a "cache", and as such, it may have side effect results (from previous and other unrelated builds), and the need for artifact availability check (this is the full name of logic you write about) is exactly that: to ensure there is no "side effect" in your build, to fail if it would fail with empty local repository.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

I strongly disagree with removing (or making it disabled by default) of this feature, especially as more advanced features (split repository, remote repository filtering) relies on availability information very much, and a healthy development environment should not, should never rely on contents of LM ("what i have in local repository"), is is just a cache, like a used napkin, with all the breadcrumbs from your breakfast and yesterday lunch. To repeat again, the ultimate check of build correctness is to nuke your local repository, and ensure it passes OK. And if it passes with empty local repository, it will pass with prepopulated one as well.

About artifact identity: yes, the identity of an artifact is not GAV, it is actually RGAV (the repository being part of it), as there is no any kind of "global artifact police" that would prevent overlapping GAVs being present in different repositories. So, talking about GAV makes sense only if you mention it's origin as well. This will become more and more prevalent, as Maven Central loose it's "central" position, that is somewhat already happening: more and more forges emerge (atlassian, redhat, spring, former JCentral, etc) with their own repositories, and actually this is good.

On related note, here is an interesting reading (and look at the comments as well): https://www.morling.dev/blog/maven-what-are-you-waiting-for/

Maven was historically a bit "hard" with projects that were using dependencies not solely on Central (so build had to contain multiple remote repositories) and this is about to change, we are working on it. For example, with Remote Repository Filter (RRF) feature you can vastly improve your build times, and make Maven be precisely directed toward needed remote repository, instead to let it "sift thru" all repositories to get the artifact. Using RRF you can Maven behave as fast as it would be only one repository used (the Central).

More reading here https://github.com/cstamas/rrf-demo

All the problems are even enlarged with use of repository manager (RM) groups (virtual repo) features, and the "recommended" Maven settings mirrorOf:*. What happens, is that this completely overwrites the origin information in your local repository, and Maven has no idea about it. Also, it silently assumes there are NO overlapping artifacts out there, but nobody talks about this aspect. What happens under the hood here, that even if your build contains definitions for "central", "atlassian", "redhat", etc, the mirrorOf setting overrides all these, and will make all these originate from your single mirrorOf repository ID (hence overlapping artifacts cannot coexist, as all of them are coalesced into single origin). In short, repository groups with recommended mirrorOf settings are to be avoided, instead let Maven be aware of your proxy and hosted repositories (enumerate them in POM), apply some solution to properly set URLs (ie. in a settings.xml profile) to make project "transportable" (ie. CI vs office workstation vs home laptop on VPN).

Usually what happens with builds like these in large(er) companies is that Maven looses information, or in other words, the "build comprehension" in Maven lacks things: a developer needs a new artifact, that alas, is not available from existing public group, so he informs the RM team about this ("add a proxy to repo X please"). The RM team, unaware of developers project adds proxy to X and stash it into public group. Next what happens that CI or developer happily builds the project, as Maven can get the new artifact from the group. But the information about the origin is lost on Maven side. Builds like these completely lack information about remote repositories they need, just take a project like that from company environment (for example company open sources it), and try to build it where there is no RM public group accessible -> is deemed to fail. To make it work, developers need to "reverse engineer" what comes from where, and make up the missing remote repository informations.

Moreover, if you consume Atlassian forge and follow their doco how to consume it, there is another problem: Atlassian actually does not publish their "hosted" repository (where their produced artifacts are hosted) but a group. Moreover, that group of theirs contains Maven Central proxy as well. Hence, in a moment you declare Atlassian repository in your build, you immediately get Maven Central present twice, and you can get slf4j-api from Maven Central but also Atlassian. Now, given the discussion about this above, they are NOT CONSIDERED SAME. For example, split repository feature would physically put them in different places (for this to happen you would need to reverse the repo order like rrf-demo project does, but still).

Here, another new feature could help to assure they are "same" (as by coordinate RGAV they are not): the "trusted checksum source", that is able to verify (even by using "strong" checksum as SHA-2) that the artifact, regardless of origin, has same checksum (so to prevent spoofing).

In short, what "availability check" does is comparing the set of project defined remote repositories and the set of origin repositories of artifact in local repository, and ensuring there is an intersection between the two. If there is none, the artifact is "not available" (created MRESOLVER-333). Effect is same, as building with empty local repository: build failure as Maven would be unable to resolve given artifact (but it has to try to be sure about it).

All that above said, there is one big (merely technical) issue about all of this: the R component of RGAV artifact identity, as it is the repository ID. Some may call a remote repository "atlassian", or "atlassian-forge" or "Atlassian-repo", etc. In controlled environment (company, or some project umbrella with several related projects) this should not be a problem, hopefully the build managers can agree on IDs and align them. So, with solution applied, this should work transparently for them. But in "wild" (ie. you build several unrelated OSS projects), there are usually always differences found, so worst happens in this case is Maven does remote access to ensure that current project should have it available.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

My overall conclusion to this discussion would be:
1. **do not stick to your local repository** (is considered transient by Maven): as part of regular developer hygiene, among many existing stuff (updating OS, updating tooling, etc), I would add "clean your local repository regularly". Make it part of your routine. Do not stick to it, it has nothing "precious" in it. If it has, and you have not setup a Maven Repository Manager already, then you are already screwed.
2. {**}help yourself, so Maven can help you as well{**}: if you are "nay" for 1st bullet, setup a local repository manager that will fix both your problems: will make initial population of local repository amazingly fast, and will also guarantee you availability as well. Do this even if you have to use company MRM (so make it company MRM <- your MRM <- your Maven). On all workstations I use  I have MRM set up with shared configuration. HOW you do this is MRM provider dependent, but you have a huge choice to choose from (OSS and "free" as well).
3. {**}avoid to use MRM groups{**}: rule of thumb defined in 2007 when group feature of Px1/Nx1 was laid down was: "groups should be consumed by clients only (Maven), but NEVER by other MRMs (as they obscure origin)". Put in mix staging repositories as well, as they are "come and go" repositories, and you have a done recipe for disaster. But today, I'd rephrase this to "avoid to use groups at all". It may sound a bit laborious, but every advanced MRM can reveal what the given URL endpoint is. Take for example ASF Nx2: the public group is this: https://repository.apache.org/content/repositories/public/.meta/repository-metadata.xml (mix of releases and snapshots repository), while staging group https://repository.apache.org/content/repositories/staging/.meta/repository-metadata.xml has very volatile member list (as staging repositories come and go). In essence, always make sure your MRM does not proxy a group (only if it must, but then make sure you apply everything you can, ie. apply some filters etc), and same thing, on Maven side in projects, always declare all your remote repositories (and again, up to you, how you solve the URLs). For example, on all my workstations, I have several set of settings.xml (for OSS work, for company/VPN work, etc but in all case my local MRM is being used). It is up to you to set up your environment properly.
